### PR TITLE
Add decryption to nc refusal

### PR DESCRIPTION
**Pull request by SteveScorfield**
# Issue Summary
https://collaborate2.ons.gov.uk/jira/browse/FWMT-2688 (only the decryption part)

# Proposed Changes
Added a decryption based class along with a processor to transform the house holder name from PGP encrypted to readable names. 

# Additional Info
Test locally with the TM mock or in WL once the environment variables have been set correctly to connect with RM Case API.

Note: This PR also includes the NC create and cancel development work that Jacob has done. 

# Checklist
- [ ] Tests
- [x] Sensitive Data not being logged
- [ ] Documentation Updated ( where required)

# How it was tested
- [ ] Unit tested
- [ ] Ran the acceptance tests
- [X] Manually Tested
- [ ] Not tested becase my coding skills are flawless

# Screenshots

### CODE REVIEW

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+  private static PGPLiteralData asLiteral(final InputStream clear) throws IOException, PGPException {
+    final PGPObjectFactory plainFact = new PGPObjectFactory(clear, new JcaKeyFingerprintCalculator());
+    final Object message = plainFact.nextObject();
+    if (message instanceof PGPCompressedData) {
+      final PGPCompressedData cData = (PGPCompressedData) message;
+      final PGPObjectFactory pgpFact = new PGPObjectFactory(cData.getDataStream(), new JcaKeyFingerprintCalculator());
+      // Find the first PGPLiteralData object
+      Object object = null;
+      for (int safety = 0; (safety++ < 1000) && !(object instanceof PGPLiteralData); object = pgpFact.nextObject()) {
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java:106]: Can we simplify this code? extract to it's own method returning the PGPLiteralData?
Also not sure the for loop is the best way to do this - what is the significance of the 1000? 
Could you use a do while loop in an separate method?

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      String errorMsg = "Failed to Authenticate with RM Case API";
+      gatewayEventManager
+          .triggerErrorEvent(this.getClass(), errorMsg, "<N/A_CASE_ID>", FAILED_TM_AUTHENTICATION);
+      throw new GatewayException(GatewayException.Fault.SYSTEM_ERROR, errorMsg, e);
+    } finally {
+      service.shutdown();
+    }
+  }
+
+  private HttpHeaders makeAuthHeader() throws GatewayException {
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java]: Looking at the code on my IDE- this method is never used? Did you mean to refactor the content of getCase and call makeAuthHeader??

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    HttpHeaders httpHeaders = new HttpHeaders();
+    if (isAuthed()) {
+      httpHeaders.setBearerAuth(auth.getAccessToken());
+    }
+    return httpHeaders;
+  }
+
+  public CaseDetailsDTO getCase(String caseId) throws GatewayException {
+    String basePathway = basePath + caseId;
+    if ((!isAuthed() || isExpired()) && !cometConfig.clientId.isEmpty() && !cometConfig.clientSecret.isEmpty())
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java:82]: duplicated in the makeAuthHeader method above - maybe use that?

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+      gatewayEventManager
+          .triggerErrorEvent(this.getClass(), errorMsg, "<N/A_CASE_ID>", FAILED_TM_AUTHENTICATION);
+      throw new GatewayException(GatewayException.Fault.SYSTEM_ERROR, errorMsg, e);
+    } finally {
+      service.shutdown();
+    }
+  }
+
+  private HttpHeaders makeAuthHeader() throws GatewayException {
+    if ((!isAuthed() || isExpired()) && !cometConfig.clientId.isEmpty() && !cometConfig.clientSecret.isEmpty())
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java]: No biggie but I'd put this entire line into it's own method returning true or false - i.e boolean needsAuth(); Makes it easier to read.

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+      if (secretKey == null) {
+        throw new GatewayException(GatewayException.Fault.SYSTEM_ERROR, "failed to find secret key for: " + encryptedFileKeyId);
+      }
+    } catch (IOException | PGPException e) {
+
+      throw new GatewayException(GatewayException.Fault.SYSTEM_ERROR, e, "Failed to obtain secret key");
+
+    }
+    try {
+      InputStream decryptedData = encryptedData.getDataStream(
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java:56]: No null check for encryptedData - if we put it inside the if block statement mentioned above then we should be gold

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    Iterator<PGPPublicKeyEncryptedData> encryptedObjects;
+    long encryptedFileKeyId = 0;
+    try {
+      householder = householder.substring(1, householder.length() - 1);
+      encryptedObjects = getEncryptedObjects(Base64.getDecoder().decode(householder));
+    } catch (IOException e) {
+
+      throw new GatewayException(GatewayException.Fault.SYSTEM_ERROR, "Failed to obtain decryption data");
+    }
+    while (encryptedObjects.hasNext()) {
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java]: do we need a loop here?  This is probably a decision point for the entire method , maybe better with an if statement - then nest the code below inside the if statement?

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/NamedHouseholderRetrieval.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+  @Autowired
+  private GatewayEventManager eventManager;
+
+  @Value("${decryption.pgp}")
+  private Resource privateKey;
+
+  @Value("${decryption.password}")
+  private String privateKeyPassword;
+
+  public String getAndSortRmRefusalCases(String caseId, CaseDetailsDTO houseHolder) throws GatewayException {
```
Comment (ocallaghanit) [inline src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/NamedHouseholderRetrieval.java:35]: This method is a great candidate for using lambda expressions.Lets go through this together?

**Review by ocallaghanit — CHANGES_REQUESTED** [review]
@SteveScorfield ping me when you have a moment - we can go through this together

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java (hunk 2 above)
Comment (rameshons) [reply src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java]: Refer CometRestClient.java constructor for resttemplate.(connection pool implemented)
Use same way in RmRestClient constructor remove rest template builder.

Comment (SteveScorfield) [reply src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java]: Wasn't needed. This has been removed.

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java (hunk 4 above)
Comment (SteveScorfield) [reply src/main/java/uk/gov/ons/census/fwmt/jobservice/http/rm/RmRestClient.java]: Wasn't needed. This has been removed.

File: src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java (hunk 5 above)
Comment (ocallaghanit) [reply src/main/java/uk/gov/ons/census/fwmt/jobservice/nc/utils/DecryptNames.java:56]: @SteveScorfield can encryptedData ever be null?

**Review by ocallaghanit — APPROVED** [review]

**Pull request merged**